### ISSUE TITLE
Postgres fails inside transaction without initialization

### ISSUE THREAD

**Issue by abaldwin88** [seamusabshere/upsert#84]
Running an upsert command inside a transaction fails unless you first run it outside a transaction first.   psql version 9.4.5 and 9.5.4
### Fails if we run inside a transaction first

```
require 'upsert/active_record_upsert'
Pet.transaction do
  Pet.upsert({name: 'Jerry'}, breed: 'beagle', created_at: Time.now, updated_at: Time.now)
end

```
### Works correctly if we run outside a transaction first to create the postgres function

```
require 'upsert/active_record_upsert'
Pet.upsert({name: 'Jerry'}, breed: 'beagle', created_at: Time.now, updated_at: Time.now)

Pet.transaction do
  Pet.upsert({name: 'Jerry'}, breed: 'beagle', created_at: Time.now, updated_at: Time.now)
end
```

See minimal example...
https://github.com/abaldwin88/upsert_minimal

Is this a known issue?  In our use case we will never be running upsert outside of a transaction.

 Is there a recommended work around?  We can add the `CREATE OR REPLACE FUNCTION` function in a migration.

**Comment by abaldwin88** [comment]
I believe the issue revolves around the manner in which a missing function is detected... https://github.com/seamusabshere/upsert/blob/1d26591ca0236aa153c1d7be4dcc8eeb2121c15e/lib/upsert/merge_function/PG_Connection.rb#L22-L27

Catching PG::Error when inside a transaction means that the transaction has already been aborted on the database side but we continue to execute commands against it.  

Not sure what a good solution would be.  Issuing a ROLLBACK or COMMIT wouldn't be appropriate.  Could possibly consider moving the exception catch into the query but that would impact performance.

**Comment by pnomolos** [comment]
@seamusabshere We could query the relevant DB and check for the existence of a function rather than checking for failure?

**Comment by pnomolos** [comment]
@abaldwin88 See #87 (once Travis checks everything off as being A-OK).  I'm working up a better solution, though, as there's a significant performance drop when using savepoints so liberally.

**Comment by abaldwin88** [comment]
@pnomolos Excellent!

Right now I am creating the function before using upsert.  Not great but our use case is only a single upsert command (for now).   

Looking forward to putting in a proper fix once your PR is merged - Thanks again.

**Comment by pnomolos** [comment]
Fixed via PR #87
